certbot-ci/certbot_integration_tests/utils/acme.py

```diff
@@ -0,0 +1,202 @@
+
+    _prepare_traefik_proxy(workspace, acme_xdist)
+    _prepare_acme_server(workspace, acme_type, acme_option, acme_xdist)
+
+    return acme_xdist
+
+
+def _construct_acme_xdist(acme_type, acme_option, nodes):
+    """Generate and return the acme_xdist dict"""
+    acme_xdist = {'challtestsrv_port': 8055}
+
+    if acme_type == 'pebble':
+        acme_xdist['directory_url'] = 'https://localhost:14000/dir'
+    else:  # boulder
+        port = 4001 if acme_option == 'v2' else 4000
+        acme_xdist['directory_url'] = 'http://localhost:{0}/directory'.format(port)
+
+    acme_xdist['acme_subnet'] = '10.77.77' if acme_type == 'boulder' else '10.30.50'
+    acme_xdist['http_port'] = {node: port for (node, port)
+                               in zip(nodes, range(5200, 5200 + len(nodes)))}
+    acme_xdist['https_port'] = {node: port for (node, port)
+                                in zip(nodes, range(5100, 5100 + len(nodes)))}
+
+    return acme_xdist
+
+
+def _construct_workspace(acme_type):
+    """Create a temporary workspace for integration tests stack"""
+    workspace = tempfile.mkdtemp()
+
+    def cleanup():
+        """Cleanup function to call that will teardown relevant dockers and their configuration."""
+        for instance in [acme_type, 'traefik']:
+            print('=> Tear down the {0} instance...'.format(instance))
+            instance_path = join(workspace, instance)
+            try:
+                if os.path.isfile(join(instance_path, 'docker-compose.yml')):
+                    _launch_command(['docker-compose', 'down'], cwd=instance_path)
+            except subprocess.CalledProcessError:
+                pass
+            print('=> Finished tear down of {0} instance.'.format(acme_type))
+
+        shutil.rmtree(workspace)
+
+    # Here with atexit we ensure that clean function is called no matter what.
+    atexit.register(cleanup)
+
+    return workspace
+
+
+def _prepare_acme_server(workspace, acme_type, acme_option, acme_xdist):
+    """Configure and launch the ACME server, Boulder or Pebble"""
+    print('=> Starting {0} instance deployment...'.format(acme_type))
+    instance_path = join(workspace, acme_type)
+    try:
+        # This loads Boulder from git, that includes a docker-compose.yml ready for production.
+        if acme_type == 'boulder':
+            _launch_command(['git', 'clone', 'https://github.com/letsencrypt/boulder',
+                             '--single-branch', '--depth=1', instance_path])
+            # Allow Boulder to ignore usual limit rate policies, useful for tests.
+            os.rename(join(instance_path, 'test/rate-limit-policies-b.yml'),
+                      join(instance_path, 'test/rate-limit-policies.yml'))
+
+        # This configure Pebble using precompiled containers.
+        if acme_type == 'pebble':
+            os.mkdir(instance_path)
+            with open(join(instance_path, 'docker-compose.yml'), 'w') as file_h:
+                file_h.write('''\
+version: '3'
```

Comment:
why is this inlined instead of in a different file? why is there no file like this already in the pebble repo?

Reply:
The inline is because I need to control whether strict mode is enabled or not. It is kind of weird to make a search and replace on an existing file. For readability sake, I prefer to have the file in place, in particular as it is very short.

There is a docker-compose.yml in Pebble repo, but it is not suitable for production. It uses the local git repo to rebuild everything from the Dockerfiles. This implies 2 minutes of recompilation while the versioned dockers are already available on Docker Hub.

Even if the docker-compose file was suitable for production, I am quite not in fond with what is done with Boulder. It clone/checkout branch master and use directly the docker-compose file set their. This is weird in my opinion: it makes your stack depends on a third party. I prefer to use my own config stack, and control exactly what is setup in Travis, or every machine a developer would use to setup integration tests.

I sticked for Boulder on the current approach, to limit the modifications. But ultimately I would like also to setup Boulder with our own docker-compose file. Also, both Boulder and Pebble should be pinned from these on-the-fly generated docker-compose file, but this for another story we discussed before.

Reply:
Well, we're already relying on a third party if we're pulling down images from docker hub. I do like having the speed up, but I worry about this getting out of sync with any changes made in the boulder/pebble repo, which is the main reason to use the docker-compose file in the repos.

Reply:
I am following the official documentation to build the config, with it we can use pinned versions like in Boulder while the one in the git repo compile master, and I need to customize several ports in the config.

Also, the concern you express is based on the complexity in Boulder. Pebble is designed to be simple. 

Can we let this as it, and modify only if it becomes a problem?

Reply:
Well, why not just fix it in the docker-compose in Pebble since we're fixing it anyway? Like Boulder does? Since the code is written anyway, it should live in the best possible place; why not have all Pebble users get the benefit of it?

Reply:
That is right, I think the docker-compose.yml that is written in my code has a better usability than the one currently in Pebble repo. I will open a PR about that. But this is a commitment for Pebble maintainers, because the docker-compose.yml file needs to be updated at each release, like in Boulder.

So we will see. If the change is accepted, then we can checkout the GIT repo and use the docker-compose.yml. But if it is not, I really would like to stay with the current approach. 

If you agree, then we could consider this conversation as resolved, either by an update when the PR is accepted, or by sticking to the current code.

Reply:
Based on the comments there, looks like that'll get into the pebble repo! So looks promising for getting into this PR even.